src/utils/cosmosdb_model_versioned_ttl.ts

```diff
@@ -0,0 +1,153 @@
+  CosmosErrors,
+  DocumentSearchKey,
+  toCosmosErrorResponse
+} from "./cosmosdb_model";
+import { asyncIterableToArray } from "./async";
+
+/**
+  The purpose of this class is to isolate the `updateTTL` logic, 
+  it should be used only for those models which does not generate a lot 
+  of versions because the logic to update the ttl is quite expensive.
+ 
+  NB: 
+  if You have more than 100 versions the update operation could not be transactional, 
+  see https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/transactional-batch?tabs=dotnet
+ */
+
+export class CosmosdbModelVersionedTTL<
+  T,
+  TN extends Readonly<T>,
+  TR extends Readonly<T & RetrievedVersionedModel>,
+  ModelIdKey extends keyof T,
+  PartitionKey extends keyof T = ModelIdKey
+> extends CosmosdbModelVersioned<T, TN, TR, ModelIdKey, PartitionKey> {
+  constructor(
+    container: Container,
+    protected readonly newVersionedItemT: t.Type<TN, ItemDefinition, unknown>,
+    protected readonly retrievedItemT: t.Type<TR, unknown, unknown>,
+    protected readonly modelIdKey: ModelIdKey,
+    protected readonly partitionKey?: PartitionKey
+  ) {
+    super(
+      container,
+      newVersionedItemT,
+      retrievedItemT,
+      modelIdKey,
+      partitionKey
+    );
+  }
+
+  /**
+    Using transactionalBatch updates the ttl field for all the versions of the document
+    identified by the searchKey.
+   */
+
+  public updateTTLForAllVersions(
+    searchKey: DocumentSearchKey<TR, ModelIdKey, PartitionKey>,
+    ttl: NonNegativeNumber
+    // eslint-disable-next-line
+  ): TE.TaskEither<CosmosErrors, TE.TaskEither<never, readonly any[]>> {
+    return pipe(
+      this.findAllVersionsBySearchKey(searchKey),
+      TE.map(RA.rights),
+      TE.map(
+        RA.map(doc => ({
+          id: doc.id,
+          operationType: "Patch" as const,
+          resourceBody: {
+            operations: [
+              {
+                /**
+                    add will replace the path with the passed value if the path alreadye exists,
+                    otherwise it will add the path
+                 */
+                op: PatchOperationType.add,
+                path: `/ttl`,
+                value: ttl
+              }
+            ]
+          }
+        }))
+      ),
+      TE.map(RA.chunksOf(100)),
+      TE.map(
+        RA.map(x =>
+          pipe(x, RA.toArray, operations =>
+            super.batch(
+              operations,
+              `${
+                searchKey.length === 1 ? searchKey[0] : searchKey[1]
+              }` as NonEmptyString
+            )
+          )
+        )
+      ),
+      TE.chainW(TE.sequenceSeqArray),
+      TE.chainW(responses =>
+        pipe(
+          responses,
+          RA.filter(response => response.code !== 200),
+          TE.fromPredicate(
+            errors => errors.length === 0,
+            _ =>
+              CosmosErrorResponse({
+                message: `Error updating ttl for ${searchKey}`,
```

Comment:
Maybe we need more information in here for example what batch fail.

Reply:
Can we add `errors` info here?

Reply:
Do we need to see all the status codes of the failed chunks? Is it useful?

Reply:
Chunks are processed sequentially, so only one can fail. The information we need is what chunk fail, to understand how must be handled the DB inconsistency (previous chunks are applied and future aren't)

Reply:
Add a comment in [here](https://github.com/pagopa/io-functions-commons/pull/296#discussion_r1000303300) that have an impact on this scenario.

Reply:
Sorry I'm loosing something. Failed batch operations with `status !== 200` don't interrupt the execution chain. IMHO a failed batch must interrupt the execution with an error.

Reply:
You'r right, even if we get a fail in one chunk (status != 200) we continue to process next chunks (if any).
The target, in my opinion, is to process as many chunks as we can. But we can move the check within the task executed sequentially, to be sure to stop as soon as we get an error.

Reply:
> process as many chunks as we can

I disagree on this, especially for the TTL logic. If a chunk that include the **last version** of the document succeded and any other not, the unrepaired inconsistency (before that the ttl logic take effect) will make active an old version of the document.

Reply:
Done in [ee1803e](https://github.com/pagopa/io-functions-commons/pull/296/commits/ee1803e704acff65bd0231a7b5a8d0dfe7088159)